File: packages/performance/ios/Plugin/FirebasePerformance.swift
import Foundation

import FirebaseCore
import FirebasePerformance

@objc public class FirebasePerformance: NSObject {
    var traces: [String: Trace] = [:]

    override init() {
        if FirebaseApp.app() == nil {
            FirebaseApp.configure()
        }
    }

    @objc public func startTrace(_ traceName: String) {
        let trace = Performance.startTrace(name: traceName)
        self.traces[traceName] = trace
    }

    @objc public func stopTrace(_ traceName: String) {
        guard let trace = self.traces[traceName] else {
            return
        }
        trace.stop()
        self.traces.removeValue(forKey: traceName)
    }

    @objc public func incrementMetric(_ traceName: String, _ metricName: String, _ incrementBy: Int) {
        guard let trace = self.traces[traceName] else {
            return
        }
        trace.incrementMetric(metricName, by: Int64(incrementBy))
    }

    @objc public func getTraceByName(_ traceName: String) -> Trace? {
        return self.traces[traceName]
    }

    @objc public func setEnabled(_ enabled: Bool) {
        Performance.sharedInstance().isDataCollectionEnabled = enabled
    }

    @objc public func isEnabled() -> Bool {
        return Performance.sharedInstance().isDataCollectionEnabled
    }

    @objc public static func putAttribute(_ trace: Trace, _ attribute: String, _ value: String) {
        trace.setValue(value, forAttribute: attribute)
    }

    @objc public static func getAttribute(_ trace: Trace, _ attribute: String) -> String? {
        return trace.value(forAttribute: attribute)
    }

    @objc public static func getAttributes(_ trace: Trace) -> [String: String] {
        return trace.attributes
    }

    @objc public static func removeAttribute(_ trace: Trace, _ attribute: String) {
        trace.removeAttribute(attribute)
    }

    @objc public static func putMetric(_ trace: Trace, _ metricName: String, _ num: Double) {
        trace.setValue(Int64(floor(num)), forMetric: metricName)
    }

    @objc public static func getMetric(_ trace: Trace, _ metricName: String) -> Int64 {
        return trace.valueForMetric(metricName)
    }

    @objc public func record(_ traceName: String, _ startTime: Double, _ duration: Double, _ attributes: [String: String], _ metrics: [String: Double]) {
        let trace = getTraceByName(traceName)
        let currentTime = Date().timeIntervalSince1970 * 1000
        let startDelay = max(0, (startTime - currentTime) / 1000)

        DispatchQueue.global().asyncAfter(deadline: .now() + startDelay) {
            for (key, value) in attributes {
                FirebasePerformance.putAttribute(trace!, key, value)
            }
            for (key, value) in metrics {
                FirebasePerformance.putMetric(trace!, key, value)
            }
            self.startTrace(traceName)

            DispatchQueue.global().asyncAfter(deadline: .now() + duration / 1000) {
                self.stopTrace(traceName)
            }
        }
    }
}

File: packages/performance/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/performance/FirebasePerformance.java
package io.capawesome.capacitorjs.plugins.firebase.performance;


import com.google.firebase.perf.metrics.Trace;
import java.util.HashMap;
import java.util.Map;

public class FirebasePerformance {

    private HashMap<String, Trace> traces = new HashMap<String, Trace>();

    public void startTrace(String traceName) {
        Trace trace = this.getFirebasePerformanceInstance().newTrace(traceName);
        trace.start();
        this.traces.put(traceName, trace);
    }

    public void stopTrace(String traceName) {
        Trace trace = this.traces.get(traceName);
        trace.stop();
        this.traces.remove(traceName);
    }

    public void incrementMetric(String traceName, String metricName, Integer incrementBy) {
        Trace trace = this.traces.get(traceName);
        trace.incrementMetric(metricName, incrementBy);
    }

    public Trace getTraceByName(String traceName) {
        if (this.traces.containsKey(traceName)) {
            return this.traces.get(traceName);
        }
        return null;
    }

    public void setEnabled(Boolean enabled) {
        this.getFirebasePerformanceInstance().setPerformanceCollectionEnabled(enabled);
    }

    public Boolean isEnabled() {
        return this.getFirebasePerformanceInstance().isPerformanceCollectionEnabled();
    }

    public static void putAttribute(Trace trace, String attribute, String value) {
        trace.putAttribute(attribute, value);
    }

    public static String getAttribute(Trace trace, String attribute) {
        return trace.getAttribute(attribute);
    }

    public static Map<String, String> getAttributes(Trace trace) {
        return trace.getAttributes();
    }

    public static void removeAttribute(Trace trace, String attribute) {
        trace.removeAttribute(attribute);
    }

    public static void putMetric(Trace trace, String metricName, long num) {
        trace.putMetric(metricName, num);
    }

    public static long getMetric(Trace trace, String metricName) {
        return trace.getLongMetric(metricName);
    }

    private com.google.firebase.perf.FirebasePerformance getFirebasePerformanceInstance() {
        return com.google.firebase.perf.FirebasePerformance.getInstance();
    }
}
